
## 1. Chapter 4 - Data Assembly

### 1.1 Tidy Data
""" The idea of "Tidy Data" comes from RStudio developer Hadley Wickham in an article by the same name, published in 2014 by the Journal of Statistical Software. Wickham defined a "framework" set of 3 characteristics that all "tidy," or easily analyzable, data share (Tidy Data, Section 2.3):

Each variable forms a column.
Each observation forms a row.
Each type of observational unit forms a table.
Wickham makes the observation that his three rules correspond to Edgar F. Codd's 3rd Normal Form (3NF) that forms the backbone of relational database design. We'll discuss relevant parts of 3NF in Section 1.3 below, and look at Tidy Data in more detail as we discuss Exploratory Data Analysis.

Wickham, H. (2014). Tidy data. Journal of Statistical Software, 59(10), 1-23."

### 1.2 Adding Rows
"Section 4.3.1 discusses adding (concatenating) rows from three dataframes with identical column names, adding a series to a dataframe as well as converting the series to a dataframe before adding, and finally, adding dataframes while ignoring the indexes so numbering is continuous."

Some examples follow. Notice we made slight changes in the file naming to be able to work on this program from the source: https://github.com/chendaniely/pandas_for_everyone/tree/master/data

### Source: https://github.com/chendaniely/pandas_for_everyone/tree/master/data

In [1]:
#Note we had to slightly change the names of the data files

import pandas as pd
df1 = pd.read_csv("concat_1.txt")
df2 = pd.read_csv("concat_2.txt")
df3 = pd.read_csv("concat_3.txt")

print(f'{df1}\n\n{df2}\n\n{df3}')

    A   B   C   D
0  a0  b0  c0  d0
1  a1  b1  c1  d1
2  a2  b2  c2  d2
3  a3  b3  c3  d3

    A   B   C   D
0  a4  b4  c4  d4
1  a5  b5  c5  d5
2  a6  b6  c6  d6
3  a7  b7  c7  d7

     A    B    C    D
0   a8   b8   c8   d8
1   a9   b9   c9   d9
2  a10  b10  c10  d10
3  a11  b11  c11  d11


In [2]:
print(pd.concat([df1, df2, df3]))

     A    B    C    D
0   a0   b0   c0   d0
1   a1   b1   c1   d1
2   a2   b2   c2   d2
3   a3   b3   c3   d3
0   a4   b4   c4   d4
1   a5   b5   c5   d5
2   a6   b6   c6   d6
3   a7   b7   c7   d7
0   a8   b8   c8   d8
1   a9   b9   c9   d9
2  a10  b10  c10  d10
3  a11  b11  c11  d11


In [3]:
series_row = pd.Series(['n1', 'n2', 'h3', 'n4'])

print(pd.concat([df1, series_row]))

     A    B    C    D    0
0   a0   b0   c0   d0  NaN
1   a1   b1   c1   d1  NaN
2   a2   b2   c2   d2  NaN
3   a3   b3   c3   d3  NaN
0  NaN  NaN  NaN  NaN   n1
1  NaN  NaN  NaN  NaN   n2
2  NaN  NaN  NaN  NaN   h3
3  NaN  NaN  NaN  NaN   n4


In [4]:
new_df = pd.DataFrame([['n1', 'n2', 'h3', 'n4']], columns = ['A','B','C','D'])
print(new_df)

    A   B   C   D
0  n1  n2  h3  n4


### "Besides the concat() function (for adding multiple things), you can also use append() to add just one object:"

In [5]:
print(df1.append(df2))

    A   B   C   D
0  a0  b0  c0  d0
1  a1  b1  c1  d1
2  a2  b2  c2  d2
3  a3  b3  c3  d3
0  a4  b4  c4  d4
1  a5  b5  c5  d5
2  a6  b6  c6  d6
3  a7  b7  c7  d7


### "Using ignore_index=True will allow the new dataframe to index properly:"

In [6]:
print(df1.append(df2, ignore_index=True))

    A   B   C   D
0  a0  b0  c0  d0
1  a1  b1  c1  d1
2  a2  b2  c2  d2
3  a3  b3  c3  d3
4  a4  b4  c4  d4
5  a5  b5  c5  d5
6  a6  b6  c6  d6
7  a7  b7  c7  d7


## 1.4 Adding Columns


In [7]:
print(f'{df1}\n\n{df2}')


    A   B   C   D
0  a0  b0  c0  d0
1  a1  b1  c1  d1
2  a2  b2  c2  d2
3  a3  b3  c3  d3

    A   B   C   D
0  a4  b4  c4  d4
1  a5  b5  c5  d5
2  a6  b6  c6  d6
3  a7  b7  c7  d7


In [8]:
df1['E'] = df2['D']
df1

A   B   C   D   E
0  a0  b0  c0  d0  d4
1  a1  b1  c1  d1  d5
2  a2  b2  c2  d2  d6
3  a3  b3  c3  d3  d7

"In most cases, that will be the easiest way to add a column.

Pandas does support the "join" method of concatenating or merging dataframes."

In [9]:
# **employees**
# emp_id | emp_name | dept_id
# --|-----|-----------
# 1 | Tom | 1
# 2 | Mary | 2
# 3 | John | 3
# 4 | Tim | 1
# 5 | Jenny | 

# NOTE: Jenny is a new hire not assigned a department yet.<br>
# **departments**

# dept_id | dept_name
# --------|----------
# 1 | HR
# 2 | Development
# 3 | Marketing
import numpy as np

# Have to put a 0 in for Jenny, then adjust it.
emp_df = pd.DataFrame({'emp_id':[1,2,3,4,5], 'emp_name':['Tom', 'Mary','John', 'Tim', 'Jenny'], 'dept_id':[1,2, 3, 1, 0] })
dept_df = pd.DataFrame({'dept_id':[1,2,3], 'dept_name':['HR','Development', 'Marketing']})

# Workaround to get a null into Jenny's dept_id
# emp_df.loc[emp_df['emp_name'] == 'Jenny', 'dept_id'] = np.nan

print(f'{emp_df}\n\n{dept_df}')

   emp_id emp_name  dept_id
0       1      Tom        1
1       2     Mary        2
2       3     John        3
3       4      Tim        1
4       5    Jenny        0

   dept_id    dept_name
0        1           HR
1        2  Development
2        3    Marketing


Outter join

In [10]:
print(pd.concat([emp_df, dept_df], join='outer', axis=1))

   emp_id emp_name  dept_id  dept_id    dept_name
0       1      Tom        1      1.0           HR
1       2     Mary        2      2.0  Development
2       3     John        3      3.0    Marketing
3       4      Tim        1      NaN          NaN
4       5    Jenny        0      NaN          NaN


Inner join

In [11]:
print(pd.concat([emp_df, dept_df],  join='inner', axis=1))

   emp_id emp_name  dept_id  dept_id    dept_name
0       1      Tom        1        1           HR
1       2     Mary        2        2  Development
2       3     John        3        3    Marketing


"However, concatenating is simply smashing two dataframes together based on index. Let's look at more of a true join with the merge() function."

## 1.5 Merging Data¶

In [12]:
emp_df.merge(dept_df, on="dept_id")

emp_id emp_name  dept_id    dept_name
0       1      Tom        1           HR
1       4      Tim        1           HR
2       2     Mary        2  Development
3       3     John        3    Marketing

"Notice how Jenny, whose department ID doesn't match in the departments table, is omitted. Inner join is the default. https://chrisalbon.com/python/data_wrangling/pandas_join_merge_dataframe/ demonstrates concatenation and merging very well."

## 2. Chapter 5 Missing Data

### 2.1 What is missing data and where does it come from?
"What
Conceptually, missing values in the data are exactly that - missing. Nothing there. Not "0" and not " " (blank space). Non-existance.

Programming languages generally have the concept of a "null" value, although a different word may be used. For example, pure Python uses the keyword "None" to express the concept.

Much of Pandas is based on the NumPy high-performance mathematical library. As such, Pandas and NumPy both use NaN -- "Not a Number" to express missing or null values.

Where
NaN or null values can creep into a data set in a variety of ways:

A value was not entered or entered incorrectly on a form.
An observation from a sensor was corrupt or missing.
Transcription error or typo.
Etc.
But wait, there's more
A much more insidious and hard to detect type of "null" comes in the form of a substitute value. Perhaps the data was stored in a database that didn't allow nulls (common) or due to a data-entry convention, or some other reason, missing data is not allowed to be null.

A prime example of this is the "Heart Disease Data Set" from the UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/datasets/Heart+Disease). This data suffers from a number of problems, including the fact that it is delimited by spaces and one observation takes multiple rows (10). Pandas can't read it in its native form, and this data set will be the subject of scrutiny in MSDS 621. For now, let's just load the file and look at one observation (one row)."

In [13]:
## notice we need to slightly adjust the code to work with the difference in file name

hrt_data = []
with open ('hungarian.data', 'r') as infile:
    for line in infile:
        hrt_data.append(line)

print(hrt_data[:10])


['1254 0 40 1 1 0 0\n', '-9 2 140 0 289 -9 -9 -9\n', '0 -9 -9 0 12 16 84 0\n', '0 0 0 0 150 18 -9 7\n', '172 86 200 110 140 86 0 0\n', '0 -9 26 20 -9 -9 -9 -9\n', '-9 -9 -9 -9 -9 -9 -9 12\n', '20 84 0 -9 -9 -9 -9 -9\n', '-9 -9 -9 -9 -9 1 1 1\n', '1 1 -9. -9. name\n']


"Data like the file above is a good example why you need to learn about your data. These -9s are the only negative numbers in the data set, but there are a lot of them. If they were allowed to stay -9s, they would drastically affect any analysis or machine learning algorithm.

Pages 109 - 116 of "Pandas for Everyone" does a good job of showing NaN values and the properties they (don't) possess as well as various ways they can sneak into perfectly respectable data sets through joins, user input, re-indexing, etc."

## 2.2 Finding and Dealing with missing data


"What is that saying? "The first step is admitting you have a problem?" "Pandas for Everyone" discusses one method of finding nulls based on column counts and data set shape. We will present a slightly different method so you can choose which method you prefes.

First, we will load in the data file:"

In [14]:
## notice we need to slightly adjust the code to work with the difference in file name once again


ebola = pd.read_csv("country_timeseries.txt")
ebola.head()

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
0    1/5/2015  289        2776.0            NaN            10030.0   
1    1/4/2015  288        2775.0            NaN             9780.0   
2    1/3/2015  287        2769.0         8166.0             9722.0   
3    1/2/2015  286           NaN         8157.0                NaN   
4  12/31/2014  284        2730.0         8115.0             9633.0   

   Cases_Nigeria  Cases_Senegal  Cases_UnitedStates  Cases_Spain  Cases_Mali  \
0            NaN            NaN                 NaN          NaN         NaN   
1            NaN            NaN                 NaN          NaN         NaN   
2            NaN            NaN                 NaN          NaN         NaN   
3            NaN            NaN                 NaN          NaN         NaN   
4            NaN            NaN                 NaN          NaN         NaN   

   Deaths_Guinea  Deaths_Liberia  Deaths_SierraLeone  Deaths_Nigeria  \
0         1786.0             NaN              2977.0             NaN   
1         1781.0             NaN              2943.0             NaN   
2         1767.0          3496.0              2915.0             NaN   
3            NaN          3496.0                 NaN             NaN   
4         1739.0          3471.0              2827.0             NaN   

   Deaths_Senegal  Deaths_UnitedStates  Deaths_Spain  Deaths_Mali  
0             NaN                  NaN           NaN          NaN  
1             NaN                  NaN           NaN          NaN  
2             NaN                  NaN           NaN          NaN  
3             NaN                  NaN           NaN          NaN  
4             NaN                  NaN           NaN          NaN

"Simply looking at the first few rows clearly shows we have NaNs.

Next, we will use the Pandas info() function to tell us about the columns:"

In [15]:
ebola.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122 entries, 0 to 121
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date                 122 non-null    object 
 1   Day                  122 non-null    int64  
 2   Cases_Guinea         93 non-null     float64
 3   Cases_Liberia        83 non-null     float64
 4   Cases_SierraLeone    87 non-null     float64
 5   Cases_Nigeria        38 non-null     float64
 6   Cases_Senegal        25 non-null     float64
 7   Cases_UnitedStates   18 non-null     float64
 8   Cases_Spain          16 non-null     float64
 9   Cases_Mali           12 non-null     float64
 10  Deaths_Guinea        92 non-null     float64
 11  Deaths_Liberia       81 non-null     float64
 12  Deaths_SierraLeone   87 non-null     float64
 13  Deaths_Nigeria       38 non-null     float64
 14  Deaths_Senegal       22 non-null     float64
 15  Deaths_UnitedStates  18 non-null     flo

"Notice the info() function tells us how many how many rows we have ( RangeIndex: 122 entries ), how many non-null values are in each column, and what type of data the column contains.

The book proposes several standard ways of handling the missing data:

Recode/replace (recoding as 0 is used as an example)
Fill forward using the last known good value
Fill backward using the next good value
Interpolation (by equally spacing values)
Dropping rows with missing values
Since much of the time we are doing our data wrangling as a precursor to building a machine learning model, we will take a quick look at scikit-learn's SimpleImputer that can be used for both numeric and categorical data, alone or as part of a preprocessing pipeline.

SimpleImputer can recognize both NaNs and substitute values such as our -9s above and replace them with a constant value or:

mean for the column
median for the column
most frequent value for the column
Let's demonstrate on the ebola dataframe we loaded earlier:"

Scikit-learn is built to use NumPy 2D matrixes of numbers, so we will make our data set, X, out of the numeric values and check a slice:



In [16]:
from sklearn.impute import SimpleImputer


In [17]:
X = ebola.iloc[:,2:].values
X[:5,:]

array([[ 2776.,    nan, 10030.,    nan,    nan,    nan,    nan,    nan,
         1786.,    nan,  2977.,    nan,    nan,    nan,    nan,    nan],
       [ 2775.,    nan,  9780.,    nan,    nan,    nan,    nan,    nan,
         1781.,    nan,  2943.,    nan,    nan,    nan,    nan,    nan],
       [ 2769.,  8166.,  9722.,    nan,    nan,    nan,    nan,    nan,
         1767.,  3496.,  2915.,    nan,    nan,    nan,    nan,    nan],
       [   nan,  8157.,    nan,    nan,    nan,    nan,    nan,    nan,
           nan,  3496.,    nan,    nan,    nan,    nan,    nan,    nan],
       [ 2730.,  8115.,  9633.,    nan,    nan,    nan,    nan,    nan,
         1739.,  3471.,  2827.,    nan,    nan,    nan,    nan,    nan]])

"Next, we will create our imputer, fit it to the data, then transform the data."

In [18]:
from sklearn.impute import SimpleImputer

In [19]:
np.set_printoptions(suppress=True) # Suppresses scientific notation for small numbers
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
imputer.fit(X)  
X_new = imputer.transform(X)
print(X_new[:5,:])

[[ 2776.          2335.3373494  10030.            16.73684211
      1.08           3.27777778     1.             3.5
   1786.          1101.20987654  2977.             6.13157895
      0.             0.83333333     0.1875         3.16666667]
 [ 2775.          2335.3373494   9780.            16.73684211
      1.08           3.27777778     1.             3.5
   1781.          1101.20987654  2943.             6.13157895
      0.             0.83333333     0.1875         3.16666667]
 [ 2769.          8166.          9722.            16.73684211
      1.08           3.27777778     1.             3.5
   1767.          3496.          2915.             6.13157895
      0.             0.83333333     0.1875         3.16666667]
 [  911.06451613  8157.          2427.36781609    16.73684211
      1.08           3.27777778     1.             3.5
    563.23913043  3496.           693.70114943     6.13157895
      0.             0.83333333     0.1875         3.16666667]
 [ 2730.          8115.         

In [20]:
import sklearn 
print (sklearn.__version__)

0.22.1


## Demonstrate several of dataframe merges from: Chrisalbon Import Modules 


## Import Modules
### data source: https://chrisalbon.com/python/data_wrangling/pandas_join_merge_dataframe/

In [21]:
import pandas as pd
from IPython.display import display
from IPython.display import Image

## Create a dataframe


In [22]:
raw_data = {
        'subject_id': ['1', '2', '3', '4', '5'],
        'first_name': ['Alex', 'Amy', 'Allen', 'Alice', 'Ayoung'], 
        'last_name': ['Anderson', 'Ackerman', 'Ali', 'Aoni', 'Atiches']}
df_a = pd.DataFrame(raw_data, columns = ['subject_id', 'first_name', 'last_name'])
df_a


subject_id first_name last_name
0          1       Alex  Anderson
1          2        Amy  Ackerman
2          3      Allen       Ali
3          4      Alice      Aoni
4          5     Ayoung   Atiches

## Create a second dataframe


In [23]:
raw_data = {
        'subject_id': ['4', '5', '6', '7', '8'],
        'first_name': ['Billy', 'Brian', 'Bran', 'Bryce', 'Betty'], 
        'last_name': ['Bonder', 'Black', 'Balwner', 'Brice', 'Btisan']}
df_b = pd.DataFrame(raw_data, columns = ['subject_id', 'first_name', 'last_name'])
df_b

subject_id first_name last_name
0          4      Billy    Bonder
1          5      Brian     Black
2          6       Bran   Balwner
3          7      Bryce     Brice
4          8      Betty    Btisan

## Create a third dataframe


In [24]:
raw_data = {
        'subject_id': ['1', '2', '3', '4', '5', '7', '8', '9', '10', '11'],
        'test_id': [51, 15, 15, 61, 16, 14, 15, 1, 61, 16]}
df_n = pd.DataFrame(raw_data, columns = ['subject_id','test_id'])
df_n

subject_id  test_id
0          1       51
1          2       15
2          3       15
3          4       61
4          5       16
5          7       14
6          8       15
7          9        1
8         10       61
9         11       16

## Join the two dataframes along rows

In [25]:
df_new = pd.concat([df_a, df_b])
df_new

subject_id first_name last_name
0          1       Alex  Anderson
1          2        Amy  Ackerman
2          3      Allen       Ali
3          4      Alice      Aoni
4          5     Ayoung   Atiches
0          4      Billy    Bonder
1          5      Brian     Black
2          6       Bran   Balwner
3          7      Bryce     Brice
4          8      Betty    Btisan

## Join the two dataframes along columns


In [26]:
pd.concat([df_a, df_b], axis=1)


subject_id first_name last_name subject_id first_name last_name
0          1       Alex  Anderson          4      Billy    Bonder
1          2        Amy  Ackerman          5      Brian     Black
2          3      Allen       Ali          6       Bran   Balwner
3          4      Alice      Aoni          7      Bryce     Brice
4          5     Ayoung   Atiches          8      Betty    Btisan

## Merge two dataframes along the subject_id value


In [27]:
pd.merge(df_new, df_n, on='subject_id')

subject_id first_name last_name  test_id
0          1       Alex  Anderson       51
1          2        Amy  Ackerman       15
2          3      Allen       Ali       15
3          4      Alice      Aoni       61
4          4      Billy    Bonder       61
5          5     Ayoung   Atiches       16
6          5      Brian     Black       16
7          7      Bryce     Brice       14
8          8      Betty    Btisan       15

## Merge two dataframes with both the left and right dataframes using the subject_id key

In [28]:
pd.merge(df_new, df_n, left_on='subject_id', right_on='subject_id')

subject_id first_name last_name  test_id
0          1       Alex  Anderson       51
1          2        Amy  Ackerman       15
2          3      Allen       Ali       15
3          4      Alice      Aoni       61
4          4      Billy    Bonder       61
5          5     Ayoung   Atiches       16
6          5      Brian     Black       16
7          7      Bryce     Brice       14
8          8      Betty    Btisan       15

## Merge with outer join

In [29]:
pd.merge(df_a, df_b, on='subject_id', how='outer')

subject_id first_name_x last_name_x first_name_y last_name_y
0          1         Alex    Anderson          NaN         NaN
1          2          Amy    Ackerman          NaN         NaN
2          3        Allen         Ali          NaN         NaN
3          4        Alice        Aoni        Billy      Bonder
4          5       Ayoung     Atiches        Brian       Black
5          6          NaN         NaN         Bran     Balwner
6          7          NaN         NaN        Bryce       Brice
7          8          NaN         NaN        Betty      Btisan

## Merge with inner join

In [30]:
pd.merge(df_a, df_b, on='subject_id', how='right')

subject_id first_name_x last_name_x first_name_y last_name_y
0          4        Alice        Aoni        Billy      Bonder
1          5       Ayoung     Atiches        Brian       Black
2          6          NaN         NaN         Bran     Balwner
3          7          NaN         NaN        Bryce       Brice
4          8          NaN         NaN        Betty      Btisan

## Merge with right join

In [31]:
pd.merge(df_a, df_b, on='subject_id', how='right')

subject_id first_name_x last_name_x first_name_y last_name_y
0          4        Alice        Aoni        Billy      Bonder
1          5       Ayoung     Atiches        Brian       Black
2          6          NaN         NaN         Bran     Balwner
3          7          NaN         NaN        Bryce       Brice
4          8          NaN         NaN        Betty      Btisan

## Merge with left join

In [32]:
pd.merge(df_a, df_b, on='subject_id', how='left')

subject_id first_name_x last_name_x first_name_y last_name_y
0          1         Alex    Anderson          NaN         NaN
1          2          Amy    Ackerman          NaN         NaN
2          3        Allen         Ali          NaN         NaN
3          4        Alice        Aoni        Billy      Bonder
4          5       Ayoung     Atiches        Brian       Black

## Merge while adding a suffix to duplicate column names

In [33]:
pd.merge(df_a, df_b, on='subject_id', how='left', suffixes=('_left', '_right'))

subject_id first_name_left last_name_left first_name_right last_name_right
0          1            Alex       Anderson              NaN             NaN
1          2             Amy       Ackerman              NaN             NaN
2          3           Allen            Ali              NaN             NaN
3          4           Alice           Aoni            Billy          Bonder
4          5          Ayoung        Atiches            Brian           Black

## Merge based on indexes


In [34]:
pd.merge(df_a, df_b, right_index=True, left_index=True)

subject_id_x first_name_x last_name_x subject_id_y first_name_y last_name_y
0            1         Alex    Anderson            4        Billy      Bonder
1            2          Amy    Ackerman            5        Brian       Black
2            3        Allen         Ali            6         Bran     Balwner
3            4        Alice        Aoni            7        Bryce       Brice
4            5       Ayoung     Atiches            8        Betty      Btisan